### PR TITLE
Adjoint Option

### PR DESCRIPTION
**Pull request by zhardy-lanl**
This PR introduces an adjoint option into the `LBSSolver`. This is primarily done as a first step towards separating response evaluations from the actual solver. The adjoint option was added to the `Options` struct, and to `LBSSolver::InputParameters` as well as routines for reorienting angles.

### Highlights

- Within `InitializeMaterials` a check is performed to determine if adjoint mode is on or off. This dictates whether `matid_to_xs_map` gets set with `SingleStateMGXS` or `AdjointMGXS`.
- When the `adjoint` option is specified via `SetOptions`, the routine explicitly looks it before going through other parameters because it is order sensitive. 
    - If the new flag is the same as the already set flag, nothing is done
    - If the flags are different and a discretization is not present, the solver has not been initialized and will be initialized in the appropriate form for an adjoint solve upon a call to the main `Initialize` routine.
    - If the flags are different and a discretization is present, the materials must be reinitialized, sources and boundary conditions cleared, and solutions zeroed out. This occurs within the `SetOptions` routine. The user is expected to pass the necessary source and boundary conditions with the adjoint option after initialization (as they are before). There is no good way to remove material sources, so one must be careful there.
- Tests updated to use the adjoint option.

### Notes

The previous adjoint solver pre-dated the executioner system. Its routine to reorient angles lived within its own execute routine. Since the executioner system, however, the `LBSSolver` have not implemented such a routine, which implies that the reorientation did not occur in the pre-existing adjoint solver. The tests passed because they were all scalar flux based, which does not get modified as a result of angle reorientation. The adjoint option implementation adds the reorientation routine to the base `LBSSolver`, where it does nothing, and then overrides it in the `DiscreteOrdinatesSolver` class. This allows for it to be called within the executioner system.

### CODE REVIEW

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc [hunk 1]
```diff
@@ -480,7 +478,110 @@ DiscreteOrdinatesSolver::SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::v
...
+      {
+        // Reorient flux moments
+        //
+        // Because flux moments are integrated angular fluxes, the
+        // angular flux and spherical harmonics must be evaluated at
+        // opposite angles in the quadrature integration. Taking advantage
+        // of the even/odd nature of the spherical harmonics, i.e.
+        // Y_{\ell,m}(-\Omega) = (-1)^\ell Y_{\ell,m}(\Omega), the flux
+        // moments must be multiplied by (-1)^\ell.
+        for (int m = 0; m < num_moments_; ++m)
```
Comment (ragusa) [inline modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:556]: In angular_quadrature_base.h, the HarmonicIndices struct has two entries: ell and m, following the standard math notation for spherical harmonics. 
Here, m is used has the running index for the mapping between the 2 indices {ell,m}(math) and the one index m (programming). Maybe m(programming) should be imom or n or something.

File: modules/linear_boltzmann_solvers/a_lbs_solver/lbs_solver.cc [hunk 2]
```diff
@@ -578,6 +581,38 @@ LBSSolver::SetOptions(const InputParameters& params)
...
+      options_.adjoint = adjoint;
+
+      // If a discretization exists, the solver has already been initialized.
+      // Reinitialize the materials to obtain the appropriate xs and clear the
+      // sources to prepare for defining the adjoint problem
+      if (discretization_)
+      {
+        // The materials are reinitialized here to ensure that the proper cross sections
+        // are available to the solver. Because an adjoint solve requires volumetric or
+        // point sources, the material-based sources are not set within the initialize routine.
```
[the remark this answers is not in the document]
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/a_lbs_solver/lbs_solver.cc:598]: Sources attached to materials. The only one that exists is `IsotropicMultiGrpSource`. What this comment is saying is that we do not go into `Material` objects, look for properties, and remove them for you.

File: modules/linear_boltzmann_solvers/a_lbs_solver/lbs_solver.h [hunk 3]
```diff
@@ -457,23 +457,30 @@ class LBSSolver : public opensn::Solver
...
+  virtual void SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::vector<int>& groupset_ids,
+                                                          Vec x,
                                                           PhiSTLOption which_phi);
 
+  /**
+   * A method for post-processing an adjoint solution.
+   *
+   * @note This does nothing for diffusion-based solvers.
+   */
+  virtual void ReorientAdjointSolution(){};
```
[the remark this answers is not in the document]
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/a_lbs_solver/lbs_solver.h:482]: There is a doc-string for the overridden method which details what happens. In documentation, if you look at `DiscreteOrdinatesdSolver::ReorientAdjointSolution` what you are saying would be there. I did not put it here since this routine in `LBSSolver` does not do that. I can still add it here if you'd like.

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc [hunk 4]
```diff
@@ -480,7 +478,110 @@ DiscreteOrdinatesSolver::SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::v
...
+
+      // Go through angles until all are paired
+      std::set<size_t> visited;
+      for (int m = 0; m < num_gs_angles; ++m)
+      {
+        // Skip if already encountered
+        if (visited.count(m) > 0) continue;
+
+        bool found = true;
+        for (int n = 0; n < num_gs_angles; ++n)
```
[the remark this answers is not in the document]
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:513]: This is a good idea. I will make the change.

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc (hunk 1 above)
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:556]: Fair enough. I will change this.

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc [hunk 5]
```diff
@@ -480,7 +478,110 @@ DiscreteOrdinatesSolver::SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::v
...
+    const auto& uk_man = groupset.psi_uk_man_;
+
+    // Build reversed angle mapping
+    std::map<int, int> reversed_angle_map;
+    if (options_.save_angular_flux)
+    {
+      const auto& omegas = groupset.quadrature_->omegas_;
+      const auto num_gs_angles = omegas.size();
+
+      // Go through angles until all are paired
```
Comment (ragusa) [inline modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:505]: angles --> directions

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc [hunk 6]
```diff
@@ -480,7 +478,110 @@ DiscreteOrdinatesSolver::SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::v
...
+    // Angular flux info
+    auto& psi = psi_new_local_[gs];
+    const auto& uk_man = groupset.psi_uk_man_;
+
+    // Build reversed angle mapping
+    std::map<int, int> reversed_angle_map;
+    if (options_.save_angular_flux)
+    {
+      const auto& omegas = groupset.quadrature_->omegas_;
+      const auto num_gs_angles = omegas.size();
```
Comment (ragusa) [inline modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:503]: num_gs_angles --> num_gs_dirs or num_gs_directions

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc [hunk 7]
```diff
@@ -480,7 +478,110 @@ DiscreteOrdinatesSolver::SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::v
...
+      std::set<size_t> visited;
+      for (int idir = 0; idir < num_gs_angles; ++idir)
+      {
+        // Skip if already encountered
+        if (visited.count(idir) > 0) continue;
+
+        bool found = true;
+        for (int jdir = 0; jdir < num_gs_angles; ++jdir)
+        {
+          // Angles are opposite if their sum is zero
```
Comment (ragusa) [inline modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:515]: Angles --> Directions

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc [hunk 8]
```diff
@@ -480,7 +478,110 @@ DiscreteOrdinatesSolver::SetPrimarySTLvectorFromMultiGSPETScVecFrom(const std::v
...
+
+    // Moment map for flux moments
+    const auto& moment_map = groupset.quadrature_->GetMomentToHarmonicsIndexMap();
+
+    // Angular flux info
+    auto& psi = psi_new_local_[gs];
+    const auto& uk_man = groupset.psi_uk_man_;
+
+    // Build reversed angle mapping
+    std::map<int, int> reversed_angle_map;
```
Comment (ragusa) [inline modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:499]: reversed_angle_map --> reversed_direction_map

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc (hunk 5 above)
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:505]: This is a less trivial change. Most of the code base calls directions angles, not the other way around... See `modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/sweep_chunks/aah_sweep_chunk.cc` for example. Thoughts?

Comment (ragusa) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:505]: This comment was specific to the commented line:

// Go through angles until all are paired
-->
// Go through directions until all are paired

Comment (ragusa) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:505]: if you are referring to angle_sets, I am willing to discuss this elsewhere. 
Here, we are really talking about psi^*(omega) = psi(-omega) and I think we would more likely say [gerund verb] the opposite direction rather than [gerund] the opposite angle.

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc (hunk 6 above)
Comment (ragusa) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:503]: num_gs_angles is a local var.

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc (hunk 7 above)
Comment (ragusa) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:515]: I refer to the comment

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc (hunk 8 above)
Comment (ragusa) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:499]: local var

File: modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc (hunk 5 above)
Comment (zhardy-lanl) [reply modules/linear_boltzmann_solvers/b_discrete_ordinates_solver/lbs_discrete_ordinates_solver.cc:505]: I am referring to the 24 instances of "angle" throughout (the sweeper loops over angles, not directions) versus the 7 instances of "direction" within as many lines of code. Using "direction" is the exception, not the rule in the code base, which is why I am hesitant to break from the existing convention. And yes, I am aware that these are all local variables. 

I tend to agree with you that maybe direction is better than angle, but I don't think we should start deciding to break from existing convention in small parts of the code via PRs separate topics. Let's create an issue and discuss further.

**Review by ragusa — APPROVED** [review]

**Review by andrsd — APPROVED** [review]

**Review by wdhawkins — APPROVED** [review]
